### ISSUE TITLE
Support more sophisticated registration of modules  by providing `ApplicationModules.merge(ApplicationModules)`

### ISSUE THREAD

**Issue by 9n** [spring-projects/spring-modulith#613]
I want to use `spring-modulith` in a cross application (cross team) way, where modules may be shared by multiple applications (don't be confused with `@Modulithic.sharedModules`).
Providing following project:
```
my-app // my-app is maintained by team A
├─ pom.xml
├─ src/main/java
│  ├─ com.example.app
│  │  ├─ Application.java
│  │  ├─ moduleA/
│  │  ├─ moduleB/
├─ Maven Dependencies
│  ├─ shared-module-foo.jar  // shared modules maintained by other teams. e.g. team B
│  │  ├─ com.example.shared.foo
│  ├─ shared-module-bar.jar
│  │  ├─ com.example.shared.bar
│  ├─ spring-boot-starter-shared-modules.jar
```
Currently, shared modules `foo` and `bar` can be registered by annotating `Application` with `@Modulithic(additionalPackages = { "com.example.shared" })` .
But I want shared modules get registered in a `auto configuration` way, instead of `additionalPackages`, because team A don't need to know the detail of shared modules and shared modules even don't share same parent package. 

I want to implement something like:
```java
//in spring-boot-starter-shared-modules.jar

@AutoConfiguration
class SharedModulesAutoConfiguration {

    @Bean
    @ConditionalOnMissingBean
    @AutoConfigureOrder(-1)
    public ApplicationModulesRuntime modulesRuntime(ApplicationRuntime runtime) {
        ThrowingSupplier<ApplicationModules> modules = () -> EXECUTOR
                .submit(() -> ApplicationModules.of(runtime.getMainApplicationClass()) //default implement
                    .merge(ApplicationModules.of("com.example.shared")))  // shared modules get registered
                .get();

        return new ApplicationModulesRuntime(modules, runtime);
    }
}
```
Moreover, shared modules should use `EXPLICITLY_ANNOTATED`  `ApplicationModuleDetectionStrategy`, while modules in `com.example.app` keep using the default one - `DIRECT_SUB_PACKAGES`.

Any suggestion?

**Comment by odrotbohm** [comment]
First of all, we have to, as you do in your description, acknowledge that, generally speaking, the arrangement you envision can be mapped to an `ApplicationModules` instance, albeit with different responsibilities. The constraints we currently impose is that the superset of root packages would have to be known from the main application class and all those packages getting the same `ApplicationModuleDetectionStrategy` applied.

That ultimately raises the bar for an additional different way of composing `ApplicationModules`, as introducing such a way will inevitably increase the complexity of the overall bootstrapping mechanism. Furthermore, we would have to make sure that the customization as envisioned is applied consistently, i.e. also in test executions. As such, an API allowing to merge `ApplicationModules` instances alone is not sufficient. While the bean definition you show would customize the runtime `ApplicationModules` instance, it would differ from the one used in `@ApplicationModuleTest`.

I guess we could fundamentally go with an `ApplicationModulesCustomizer` SPI loaded via `SpringFactories` exposing a `ApplicationModules customize(ApplicationModules modules)` to then add the additional modules to the instance to operate on.

That said, there are a couple of questions that we would need to find answers for:
* How does the augmentation affect the caching of `ApplicationModules`? Currently, the instances are cached by base package or root type (which contains the additional packages information). In a world in which we dynamically collect extensions of `ApplicationModules`, how do we properly cache the augmented instances?
* `ApplicationModules` currently contains fields that cannot easily be augmented. For example, `JavaClasses` cannot easily be created or used in a way that they do not change the state of the original instance.
* What about other parts of the setup that can be tweaked on `@Modulithic` such as the shared modules.

I guess we could somehow augment the list of base packages in the first step of the bootstrap by loading implementations of a `SpringFactories`-based SPI contributing additional packages. We could then treat those as if they had been declared on the annotation. That, however, would still require the same `ApplicationModuleDetectionStrategy` being used.

1.3 introduces `ApplicationModuleSource` that looks like something that could be of help here, but we'd need to first scan the classes of the contributed packages.

Long story short: I am working on a prototype as I write this, and it looks promising as of now.

**Comment by odrotbohm** [comment]
I've pushed a new `ApplicationModuleSourceFactory` SPI that can be used to contribute additional packages to detect modules in (including a customizable `ApplicationModuleDetectionStrategy` for that)  or explicitly list additional modules from. Find more information in the updated reference documentation. Any feedback appreciated!
